input 'week' may display the wrong date when the first week or last week of a year is selected

lilyspiniolas opened the pull request:
#### 0bb11f0f467f2a3476f670ce1a913fcd5f204672
<pre>
input &apos;week&apos; may display the wrong date when the first week or last week of a year is selected
<a href="https://bugs.webkit.org/show_bug.cgi?id=279276">https://bugs.webkit.org/show_bug.cgi?id=279276</a>
<a href="https://rdar.apple.com/132954506">rdar://132954506</a>

Reviewed by Aditya Keerthi.

The &apos;week&apos; input could show the wrong week number and/or year number depending on the first weekday of the selected year and the number of weeks in that year according to ISO-8601. A related issue could also cause the wrong week to appear as selected when reopening the calendar view after selecting a week depending on the user&apos;s locale, timezone, and the number of ISO-8601 weeks in that year.

The first issue was resolved by constructing a displayed string which directly uses DateComponents rather than using NSDateFormatter.

The second issue was resolved by updating the formatter used to calculate the initial value for input &quot;week&quot; to interpret the parsed date as having a time of 00:00 in the user&apos;s own timezone, preventing cases where timezone adjustments could cause the date to be interpreted as the day previous to the correct one.

* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::formatDateTime):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker iso8601DateFormatterForCalendarView]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):

Canonical link: <a href="https://commits.webkit.org/283427@main">https://commits.webkit.org/283427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2572a98b3f084100e34de28717ecf22d0c1ed18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45488 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70147 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16725 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68233 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53287 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17006 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53064 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11646 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69182 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41964 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57237 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33697 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38635 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14615 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15601 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60531 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14961 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71849 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10070 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14374 "Found 1 new test failure: http/wpt/mediastream/transfer-mediastreamtrack-to-worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60378 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10102 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57299 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60670 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14607 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8317 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1956 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41296 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42372 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43555 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42116 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -90,11 +90,17 @@
+    if (type == DateComponentsType::Week) {
+#if ENABLE(INPUT_TYPE_WEEK_PICKER)
+        // NSDateFormatter is not used here because it handles week numbering differently than ISO-8601.
+        return [NSString stringWithFormat:@"%@ %i, %i", (NSString *)inputWeekLabel(), dateComponents.week(), dateComponents.fullYear()];
```
If we're going with this approach, it's probably better to make the localized string itself be a format string.

aprotyas replied:
Agreed

lilyspiniolas replied:
Fixed in latest commit

aprotyas commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -90,11 +90,17 @@
+        // NSDateFormatter is not used here because it handles week numbering differently than ISO-8601.
+        return [NSString stringWithFormat:@"%@ %i, %i", (NSString *)inputWeekLabel(), dateComponents.week(), dateComponents.fullYear()];
+#else
+        // "week" type not supported.
```
I think our conditionalizing on `ENABLE(INPUT_TYPE_WEEK_PICKER)` makes the code clear enough to not need to say this.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -121,7 +121,7 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
     case DateComponentsType::Week:
 #if ENABLE(INPUT_TYPE_WEEK_PICKER)
-        [dateFormatter setDateFormat:[NSString stringWithFormat:@"'%@' w',' yyyy", (NSString *)inputWeekLabel()]];
+        [dateFormatter setDateFormat:[NSString stringWithFormat:@"'%@' w',' yyyy", (NSString *)weekLabel()]];
```
(remaining issue is that the width calculation isn't quite right, as we've been talking about on Slack)

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -136,6 +131,31 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
 {
+#if ENABLE(INPUT_TYPE_WEEK_PICKER)
+    if (type == DateComponentsType::Week) {
+        float numLengths[10];
```
We should use `std::array`, which is safer.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -136,6 +131,31 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
+            weekString = adoptNS([NSString stringWithFormat:@"%d%d%d%d-W50", widestNumNonZero, widestNum, widestNum, widestNum]);
+        else
+            weekString = adoptNS([NSString stringWithFormat:@"%d%d%d%d-W%d%d", widestNumNonZero, widestNum, widestNum, widestNum, widestNumOneThroughFour, widestNum]);
+        auto components = DateComponents::fromParsingWeek((String)weekString.get());
```
Is `(String)` necessary?

lilyspiniolas replied:
I believe so; fromParsingWeek requires a StringView which can be implicitly casted from Strings.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -136,6 +131,31 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
+        else
+            weekString = adoptNS([NSString stringWithFormat:@"%d%d%d%d-W%d%d", widestNumNonZero, widestNum, widestNum, widestNum, widestNumOneThroughFour, widestNum]);
+        auto components = DateComponents::fromParsingWeek((String)weekString.get());
+        if (components.has_value())
```
You can move the variable assignment above directly into the `if (...)`

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -136,6 +131,31 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
+
+        RetainPtr<NSString> weekString;
+        // W50 is an edge case here; without this check, a suboptimal choice would be made when 5 and 0 are both large and all other numbers are narrow.
+        if (numLengths[5] + numLengths[0] > numLengths[widestNumOneThroughFour] + numLengths[widestNum])
```
What about `W51`/`W52`?

lilyspiniolas replied:
If there was a case where W51 or W52 were the largest, W15 or W25 would be chosen which would have identical widths. W50 is special because it's mirrored version, W05, would display as "Week 5" instead of "Week 05" thus making it narrower

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -136,6 +131,31 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
 // We assume no formatter option shows that information.
 float LocalizedDateCache::calculateMaximumWidth(DateComponentsType type, const MeasureTextClient& measurer)
 {
+#if ENABLE(INPUT_TYPE_WEEK_PICKER)
```
Let's factor out this logic into a separate method.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -280,9 +279,12 @@ - (void)setWeekPickerToInitialValue
-    RetainPtr<NSDateComponents> dateComponents = [[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] components:unitFlags fromDate:parsedDate.get() ? parsedDate.get() : [NSDate date]];
+    RetainPtr parsedDate = [[self iso8601DateFormatterForCalendarView] dateFromString:[self _sanitizeInputValueForFormatter:_initialValue.get()]];
+
+    if (!parsedDate.get() || ![[_calendarView availableDateRange] containsDate:parsedDate.get()])
```
```suggestion
    if (!parsedDate || ![[_calendarView availableDateRange] containsDate:parsedDate.get()])
```

lilyspiniolas committed:
input 'week' may display the wrong date when the first week or last week of a year is selected
https://bugs.webkit.org/show_bug.cgi?id=279276
rdar://132954506

Reviewed by Aditya Keerthi.

The 'week' input could show the wrong week number and/or year number depending on the first weekday of the selected year and the number of weeks in that year according to ISO-8601. A related issue could also cause the wrong week to appear as selected when reopening the calendar view after selecting a week depending on the user's locale, timezone, and the number of ISO-8601 weeks in that year.

The first issue was resolved by constructing a displayed string which directly uses DateComponents rather than using NSDateFormatter.

The second issue was resolved by updating the formatter used to calculate the initial value for input "week" to interpret the parsed date as having a time of 00:00 in the user's own timezone, preventing cases where timezone adjustments could cause the date to be interpreted as the day previous to the correct one.

* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::formatDateTime):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker iso8601DateFormatterForCalendarView]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):

Canonical link: https://commits.webkit.org/283427@main
